File: sdk/kotlin/idenqa/src/main/kotlin/dev/idenqa/sdk/TemporaryFiles.kt
package dev.idenqa.sdk

import android.content.Context
import java.io.File
import java.time.Instant

/** Opaque handle to one staged temporary evidence file. */
data class CaptureTemporaryFile(val reference: String, val byteCount: Int, val createdAt: Instant)

/**
 * Bounded, app-private staging storage for in-flight evidence.
 *
 * Implementations must keep files inside app-private storage, enforce the
 * configured bounds, never write evidence bytes to logs, and remove staged
 * files deterministically. Raw evidence is never persisted across process
 * death: staging only covers an in-flight attempt.
 */
interface CaptureTemporaryFileStore {
    suspend fun stage(artifact: CapturedArtifact, category: String, now: Instant): CaptureTemporaryFile
    suspend fun data(file: CaptureTemporaryFile): ByteArray
    suspend fun remove(file: CaptureTemporaryFile)
    suspend fun purge(olderThan: Instant): Int
    suspend fun purgeAll(): Int
    suspend fun count(): Int
    suspend fun totalBytes(): Int
}

/**
 * App-private staging store backed by `Context.noBackupFilesDir`.
 *
 * Internal storage is already app-private; on API 26+ the directory is also
 * excluded from backup and device transfer. Staged files never leave the
 * private container and are deleted on every completion path.
 */
class AndroidTemporaryFileStore(
    private val root: File,
    private val maximumFileBytes: Int,
    private val maximumTotalBytes: Int,
    namespace: String,
) : CaptureTemporaryFileStore {
    constructor(
        context: Context,
        namespace: String,
        maximumFileBytes: Int,
        maximumTotalBytes: Int,
    ) : this(
        File(File(context.applicationContext.noBackupFilesDir, "idenqa-capture"), namespace),
        maximumFileBytes,
        maximumTotalBytes,
        namespace,
    )

    private val lock = Any()

    init {
        if (!CaptureConfiguration.validOpaqueReference(namespace, 128) ||
            maximumFileBytes <= 0 || maximumTotalBytes < maximumFileBytes
        ) throw IdenqaException.InvalidConfiguration
    }

    override suspend fun stage(artifact: CapturedArtifact, category: String, now: Instant): CaptureTemporaryFile = synchronized(lock) {
        if (!CaptureConfiguration.validOpaqueReference(category, 64)) throw IdenqaException.InvalidConfiguration
        if (artifact.contentType != "image/jpeg" && artifact.contentType != "image/png") throw IdenqaException.InvalidConfiguration
        val bytes = artifact.bytes()
        if (bytes.isEmpty() || bytes.size > maximumFileBytes) throw IdenqaException.TemporaryStorage
        ensureDirectory()
        if (totalBytesLocked() + bytes.size > maximumTotalBytes) throw IdenqaException.TemporaryStorage
        val file = File(root, "$category-${java.util.UUID.randomUUID()}.bin")
        try {
            file.outputStream().use { it.write(bytes) }
        } catch (_: Exception) {
            file.delete()
            throw IdenqaException.TemporaryStorage
        } finally {
            bytes.fill(0)
        }
        CaptureTemporaryFile(file.name, artifact.bytes().size, now)
    }

    override suspend fun data(file: CaptureTemporaryFile): ByteArray = synchronized(lock) {
        try {
            resolve(file).readBytes()
        } catch (_: Exception) {
            throw IdenqaException.TemporaryStorage
        }
    }

    override suspend fun remove(file: CaptureTemporaryFile) {
        synchronized(lock) { resolve(file).delete() }
    }

    override suspend fun purge(olderThan: Instant): Int = synchronized(lock) {
        var removed = 0
        root.listFiles()?.forEach { file ->
            if (file.lastModified() < olderThan.toEpochMilli()) {
                if (file.delete()) removed++
            }
        }
        removed
    }

    override suspend fun purgeAll(): Int = synchronized(lock) {
        var removed = 0
        root.listFiles()?.forEach { file -> if (file.delete()) removed++ }
        removed
    }

    override suspend fun count(): Int = synchronized(lock) { root.listFiles()?.size ?: 0 }

    override suspend fun totalBytes(): Int = synchronized(lock) { totalBytesLocked() }

    private fun totalBytesLocked(): Int = root.listFiles()?.sumOf { it.length().toInt() } ?: 0

    private fun ensureDirectory() {
        if (root.isDirectory) return
        if (root.exists() || !root.mkdirs()) throw IdenqaException.TemporaryStorage
    }

    private fun resolve(file: CaptureTemporaryFile): File {
        if (file.reference.isEmpty() || file.reference.contains('/') || file.reference != File(file.reference).name) {
            throw IdenqaException.InvalidConfiguration
        }
        return File(root, file.reference)
    }
}

File: sdk/swift/Sources/Idenqa/TemporaryFiles.swift
import Foundation

/// Opaque handle to one staged temporary evidence file.
public struct CaptureTemporaryFile: Equatable, Sendable {
    public let reference: String
    public let byteCount: Int
    public let createdAt: Date

    public init(reference: String, byteCount: Int, createdAt: Date) {
        self.reference = reference
        self.byteCount = byteCount
        self.createdAt = createdAt
    }
}

/// Bounded, app-private staging storage for in-flight evidence.
///
/// Implementations must keep files inside app-private storage, enforce the
/// configured bounds, never write evidence bytes to logs, and remove staged
/// files deterministically. Raw evidence is never persisted across process
/// death: staging only covers an in-flight attempt.
public protocol CaptureTemporaryFileStore: Sendable {
    func stage(_ artifact: CapturedArtifact, category: String, now: Date) async throws -> CaptureTemporaryFile
    func data(for file: CaptureTemporaryFile) async throws -> Data
    func remove(_ file: CaptureTemporaryFile) async throws
    func purge(olderThan: Date) async throws -> Int
    func purgeAll() async throws -> Int
    func count() async throws -> Int
    func totalBytes() async throws -> Int
}

/// FileManager-backed app-private staging store.
///
/// On iOS staged files use complete file protection where the platform
/// supports it. On every platform the directory is created inside the app's
/// private caches container with owner-only permissions.
public actor FileManagerCaptureTemporaryFileStore: CaptureTemporaryFileStore {
    private let directory: URL
    private let maximumFileBytes: Int
    private let maximumTotalBytes: Int
    private let fileManager: FileManager

    public init(
        namespace: String,
        maximumFileBytes: Int,
        maximumTotalBytes: Int,
        directory: URL? = nil,
        fileManager: FileManager = .default
    ) throws {
        guard CaptureConfiguration.validOpaqueReference(namespace, maximum: 128),
              maximumFileBytes > 0, maximumTotalBytes >= maximumFileBytes else {
            throw IdenqaError.invalidConfiguration
        }
        let root = directory ?? fileManager.urls(for: .cachesDirectory, in: .userDomainMask)[0]
            .appending(path: "dev.idenqa.capture", directoryHint: .isDirectory)
        self.directory = root.appending(path: namespace, directoryHint: .isDirectory)
        self.maximumFileBytes = maximumFileBytes
        self.maximumTotalBytes = maximumTotalBytes
        self.fileManager = fileManager
    }

    public func stage(_ artifact: CapturedArtifact, category: String, now: Date) async throws -> CaptureTemporaryFile {
        guard CaptureConfiguration.validOpaqueReference(category, maximum: 64) else {
            throw IdenqaError.invalidConfiguration
        }
        guard artifact.contentType == "image/jpeg" || artifact.contentType == "image/png" else {
            throw IdenqaError.invalidConfiguration
        }
        guard !artifact.bytes.isEmpty, artifact.bytes.count <= maximumFileBytes else {
            throw IdenqaError.temporaryStorage
        }
        try ensureDirectory()
        let existing = try await totalBytes()
        guard existing + artifact.bytes.count <= maximumTotalBytes else { throw IdenqaError.temporaryStorage }
        let reference = "\(category)-\(UUID().uuidString).bin"
        let url = directory.appending(path: reference)
        var options: Data.WritingOptions = [.atomic]
        #if os(iOS)
        options.insert(.completeFileProtection)
        #endif
        try artifact.bytes.write(to: url, options: options)
        #if os(iOS) || os(macOS)
        try? fileManager.setAttributes([.posixPermissions: 0o600], ofItemAtPath: url.path)
        #endif
        return CaptureTemporaryFile(reference: reference, byteCount: artifact.bytes.count, createdAt: now)
    }

    public func data(for file: CaptureTemporaryFile) async throws -> Data {
        try Data(contentsOf: url(for: file))
    }

    public func remove(_ file: CaptureTemporaryFile) async throws {
        let url = try url(for: file)
        guard fileManager.fileExists(atPath: url.path) else { return }
        try fileManager.removeItem(at: url)
    }

    public func purge(olderThan date: Date) async throws -> Int {
        var removed = 0
        for entry in try contents() where entry.modified < date {
            try? fileManager.removeItem(at: entry.url)
            removed += 1
        }
        return removed
    }

    public func purgeAll() async throws -> Int {
        var removed = 0
        for entry in try contents() {
            try? fileManager.removeItem(at: entry.url)
            removed += 1
        }
        return removed
    }

    public func count() async throws -> Int {
        try contents().count
    }

    public func totalBytes() async throws -> Int {
        try contents().reduce(0) { $0 + $1.size }
    }

    private func ensureDirectory() throws {
        var isDirectory: ObjCBool = false
        if fileManager.fileExists(atPath: directory.path, isDirectory: &isDirectory) {
            guard isDirectory.boolValue else { throw IdenqaError.temporaryStorage }
            return
        }
        try fileManager.createDirectory(
            at: directory,
            withIntermediateDirectories: true,
            attributes: [.posixPermissions: 0o700]
        )
    }

    private func url(for file: CaptureTemporaryFile) throws -> URL {
        guard !file.reference.isEmpty,
              file.reference == URL(fileURLWithPath: file.reference).lastPathComponent,
              !file.reference.contains("/") else {
            throw IdenqaError.invalidConfiguration
        }
        return directory.appending(path: file.reference)
    }

    private func contents() throws -> [Entry] {
        guard fileManager.fileExists(atPath: directory.path) else { return [] }
        let urls = try fileManager.contentsOfDirectory(
            at: directory,
            includingPropertiesForKeys: [.fileSizeKey, .contentModificationDateKey],
            options: [.skipsHiddenFiles]
        )
        return try urls.map { url in
            let values = try url.resourceValues(forKeys: [.fileSizeKey, .contentModificationDateKey])
            return Entry(url: url, size: values.fileSize ?? 0, modified: values.contentModificationDate ?? .distantPast)
        }
    }

    private struct Entry {
        let url: URL
        let size: Int
        let modified: Date
    }
}
